### ISSUE TITLE
x/tools/go/ssa: panic: variadic parameter must be of unnamed slice type

### ISSUE THREAD

**Issue by visualfc** [golang/go#54946]
### What version of Go are you using (`go version`)?

<pre>
$ go version
go version go1.18.4 darwin/amd64
</pre>

### What operating system and processor architecture are you using (`go env`)?

<details><summary><code>go env</code> Output</summary><br><pre>
$ go env

</pre></details>

### What did you do?

Input program: $GOROOT/test/typeparam/issue376214.go
```
package main

func add[S ~string | ~[]byte](buf *[]byte, s S) {
	*buf = append(*buf, s...)
}

func main() {
	var buf []byte
	add(&buf, "foo")
	add(&buf, []byte("bar"))
	if string(buf) != "foobar" {
		panic("got " + string(buf))
	}
}
```
I run it with `ssadump -build=G issue376214.go`

### What did you expect to see?

I expect it pass.

### What did you see instead?

```
panic: variadic parameter must be of unnamed slice type

goroutine 1 [running]:
go/types.NewSignatureType(0x0, {0x0, 0x0, 0x0}, {0x0, 0x0, 0x0}, 0xc00000d908, 0xc00000d758, 0x1)
	$home/myproj/go/src/go/types/signature.go:53 +0x29e
golang.org/x/tools/internal/typeparams.NewSignatureType(...)
	$home/myproj/tools/internal/typeparams/typeparams_go118.go:55
golang.org/x/tools/go/ssa.(*subster).signature(0x1286b00?, 0xc0002de200)
	$home/myproj/tools/go/ssa/subst.go:372 +0xd2
....
```

**Comment by mknyszek** [comment]
CC @golang/tools-team

**Comment by findleyr** [comment]
CC @timothy-king @adonovan @zpavlinovic

**Comment by timothy-king** [comment]
I was not able to reproduce this yet. Steps I tried below.

So far the root cause of similar issues has been compiling with a too old version of x/tools. Do you know what version of go you compiled ssadump with and what version of x/tools it was?

Here are the steps I took to reproduce.
```
% go install golang.org/dl/go1.18.4@latest
% ~/go/bin/go1.18.4 download
% go1.18.4 install golang.org/x/tools/cmd/ssadump@latest
% go version -m /Users/taking/go/bin/ssadump
/Users/taking/go/bin/ssadump: go1.18.4
        path    golang.org/x/tools/cmd/ssadump
        mod     golang.org/x/tools      v0.1.12 h1:VveCTK38A2rkS8ZqFY25HIDFscX5X9OoEhJd3quQmXU=
        dep     golang.org/x/mod        v0.6.0-dev.0.20220419223038-86c51ed26bb4        h1:6zppjxzCulZykYSLyVDYbneBfbaBIQPYMevg0bEwv2s=
        dep     golang.org/x/sys        v0.0.0-20220722155257-8c9f86f7a55f      h1:v4INt8xihDGvnrfjMDVXGxw9wrfxYyCjk0KbXjhR55s=
...
% /Users/taking/go/bin/ssadump ~/sdk/go1.18.4/test/typeparam/issue376214.go
% # works fine 
```

**Comment by visualfc** [comment]
> ssadump ~/sdk/go1.18.4/test/typeparam/issue376214.go

`ssadump -build=FG ~/sdk/go1.18.4/test/typeparam/issue376214.go`

**Comment by findleyr** [comment]
@visualfc please share the output of `go version -m $(which ssadump)`.

**Comment by timothy-king** [comment]
Thanks. I was missing `-build=G`. `ssadump -build=G ~/sdk/go1.18.4/test/typeparam/issue376214.go` now reproduces.

**Comment by visualfc** [comment]
> @visualfc please share the output of `go version -m $(which ssadump)`.
```
go1.18.4
	path	golang.org/x/tools/cmd/ssadump
	mod	golang.org/x/tools	v0.1.12	h1:VveCTK38A2rkS8ZqFY25HIDFscX5X9OoEhJd3quQmXU=
	dep	golang.org/x/mod	v0.6.0-dev.0.20220419223038-86c51ed26bb4	h1:6zppjxzCulZykYSLyVDYbneBfbaBIQPYMevg0bEwv2s=
	dep	golang.org/x/sys	v0.0.0-20220722155257-8c9f86f7a55f	h1:v4INt8xihDGvnrfjMDVXGxw9wrfxYyCjk0KbXjhR55s=
	build	-compiler=gc
	build	CGO_ENABLED=1
	build	CGO_CFLAGS="-g -O2"
	build	CGO_CPPFLAGS=
	build	CGO_CXXFLAGS="-g -O2"
	build	CGO_LDFLAGS="-g -O2"
	build	GOARCH=amd64
	build	GOOS=darwin
	build	GOAMD64=v1
```

**Comment by timothy-king** [comment]
The example in question:
```
func add[S ~string | ~[]byte](buf *[]byte, s S) {
	*buf = append(*buf, s...)
}
```
The issue comes from substitution of S to string. The type of `append` needs to go from a `func([]byte, S...) []byte` to a `func([]byte, string...)[]byte`.

I don't believe we can create the signature `func([]byte, string...)[]byte` from `go/types` ATM. This can either be solved by `go/types` being more permissive for this as a special case, offering an interface that lets this Signature be created, or ssa instantiating a different type as a special case, like `func([]byte, []byte...)[]byte`, and adding conversion, like treating this as `append(*buf, []byte(s)...)`.

**Comment by findleyr** [comment]
Filed #55030 for discussion of what to do about this in go/types.

**Comment by timothy-king** [comment]
Fixed at tip. Waiting for https://github.com/golang/go/commit/a366ed5982d674d7e468ccc5f8b6c5b8a402a00a to be released with 1.19.2 and https://github.com/golang/go/commit/e40a130cc02cca954ff8e33aa9f5950ecab99e3b to be released with 1.18.7 for this to be closed.

**Comment by timothy-king** [comment]
Fixes have been back-ported.
